probe-rs-tools/src/bin/probe-rs/cmd/dap_server/server/logger.rs

```diff
@@ -0,0 +1,224 @@
+use crate::cmd::dap_server::{
+    debug_adapter::{dap::adapter::DebugAdapter, protocol::ProtocolAdapter},
+    DebuggerError,
+};
+use anyhow::anyhow;
+use std::{
+    fs::File,
+    io::{stderr, BufRead, BufReader, LineWriter, SeekFrom, Write},
+    path::Path,
+    sync::Mutex,
+};
+use tempfile::tempfile;
+use tracing::{level_filters::LevelFilter, subscriber::DefaultGuard};
+use tracing_subscriber::{
+    fmt::{format::FmtSpan, writer::BoxMakeWriter, MakeWriter},
+    prelude::__tracing_subscriber_SubscriberExt,
+    util::SubscriberInitExt,
+    EnvFilter, Layer,
+};
+
+/// DebugLogger manages the temporary file that is used to store the tracing messages that are generated during the DAP sessions.
+/// For portions of the Debugger lifetime where no DAP session is active, the tracing messages are sent to `stderr`.
+pub(crate) struct DebugLogger {
+    /// This is a temporary file, created inside of `std::env::temp_dir()`,
+    /// and will be automatically removed by the OS when the last handle to it is closed.
+    ///  - When the DAP server is running, the tracing messages are sent to the console.
+    ///  - When the DAP server exits, the remaining messages in this buffer are sent to `stderr`.
+    buffer_file: Mutex<File>,
```

Comment:
Why does this have to be inside a mutex, if we're just cloning the file descriptor? We could save a `lock`, a `try_clone` and a bunch of error handling if this was just a `File` and if `buffer_file_handle` returned a `&File`.

Reply:
The buffer file is used by the `tracing` implementation to write the logs to, as well as by the `DebugLogger` for both writing and reading.  I need the lock so that I can write to the file, and then update the read pointer to make sure a future `flush()` knows where the last 'un-read' data was.

Reply:
> I need the lock

But you don't! The only `.lock()` here is when you clone the FD and the lock is immediately released. What am I not seeing?

Reply:
That was not the intention. The intention was to hold onto the lock until I've finished updating the read pointers, but I see now that it is being released. Let me work with it a bit more before we merge.

Reply:
A few things:
 - I think we should prefer `parking_lot::Mutex`, that does not return an error from lock. I'm not convinced we need to deal with poisoning here.
 - If you need the lock to stick for a bit, you can return `impl Deref<Target = File>`. That way the user (even if it's just us) doesn't have to know about the lock's internals, but the lock will be held as long as it's needed.

Reply:
@bugadani I've made the changes to hold onto the `MutexGuard` until after I've updated the file pointers.  I've done some manual testing with excessive amounts of log data being written to the Debug Console, and all seems to be in order. That said, I'd really appreciate if you can do a sanity check for me.